Add a command hook to cleanup built images

andytson-inviqa opened the pull request:
Replacing potential for discrepancy of deleting images based on matching hardcoding

andytson-inviqa committed:
Add a command hook to cleanup built images

Replacing potential for discrepancy of deleting images based on matching hardcoding

kierenevans commented:
Okay to include https://github.com/inviqa/harness-base-php/pull/515's additions?

andytson-inviqa commented:
can you tell me why they wouldn't be covered by this? (maybe in slack as it'll be a bit of discussion)

kierenevans commented:
Though it might do already actually!

Sorry for the build failure, trying to replace jenkins workers and had to cancel it

kierenevans commented on src/_base/harness/config/cleanup.yml:
```diff
@@ -0,0 +1,19 @@
+    IMAGES: = built_images(docker_service_images())
+  exec: |
+    #!php
+    docker rm $IMAGES
```
```suggestion
    docker image rm $IMAGES
```

kierenevans commented on src/_base/harness/config/cleanup.yml:
```diff
@@ -0,0 +1,19 @@
+  env:
+    IMAGES: = built_images(docker_service_images())
+  exec: |
+    #!php
```
```suggestion
    #!bash
```

andytson-inviqa committed:
Prevent shell/option injection in IMAGES

kierenevans commented:
Testing this locally with `./build && ./test php static`, I'm left with a 922MB image:
```bash
docker image ls | head -n30
REPOSITORY                            TAG                        IMAGE ID       CREATED         SIZE
<none>                                <none>                     49581cfebaeb   2 minutes ago   922MB
```
Seems to be due to the multi-stage build for php-fpm from console:
```bash
■ > docker-compose build php-fpm
Building php-fpm
Step 1/13 : FROM ci-php-sample-static:b1908f9e057628fc039b085c6ee94d9b659a5dfb-console as console
 ---> 91b3f0326547
Step 2/13 : RUN if [ -d /app/tools/assets/ ]; then rm -rf /app/tools/assets/; fi
 ---> Running in aad57c2336b5
Removing intermediate container aad57c2336b5
 ---> 442cfdabf896
Step 3/13 : RUN if [ -d /app/node_modules/ ]; then rm -rf /app/node_modules/; fi
 ---> Running in 3cd1e964ed86
Removing intermediate container 3cd1e964ed86
 ---> 49581cfebaeb

Step 4/13 : FROM my127/php:8.0-fpm-buster
 ---> af838a7410dc
Step 5/13 : WORKDIR /app
 ---> Running in 18b4e8e15196
Removing intermediate container 18b4e8e15196
 ---> e07650a89137
 ```

andytson-inviqa commented:
that's why I'm thinking `docker image prune` in CI, either here or a regular cron. Can also alternatively label the image to delete it by filter, but would need to remember that practice for any other multi-stage build (which may not even be FROM a project's image)

kierenevans commented on src/_base/harness/config/cleanup.yml:
```diff
@@ -0,0 +1,20 @@
+  exec: |
+    #!bash
+    IMAGES=($IMAGES)
+    docker image rm --force -- "${IMAGES[@]}"
```
This seems to catch the stragglers without needing to prune, though it deletes nginx and php-fpm early and then complains they don't exist later:
```suggestion
    for image in ${IMAGES[@]}; do
      docker images --filter=since="${image}" -q | xargs --no-run-if-empty docker image rm --force
    done
    docker image rm --force -- "${IMAGES[@]}"
```

This is currently running in dynamic mode too so need to drop `--no-run-if-empty` if on macOS at the risk of an error occurring if no entries:

```suggestion
    for image in ${IMAGES[@]}; do
      docker images --filter=since="${image}" -q | xargs docker image rm --force
    done
    docker image rm --force -- "${IMAGES[@]}"
```

andytson-inviqa replied:
I think a prune is fine, it only deletes dangling images (of which would only be multi-stage builds and previous build failures not cleaned up) anyway

andytson-inviqa replied:
only the LABEL way tho would be suitable for concurrent executors, both filter=since and prune would have issues

andytson-inviqa replied:
I've added the LABEL way as is good for local too, but I think if there's good enough reason for anything more then for CI it be prune as is simple with the same consequences as complex filter=since, but even better because it cleans up all prior dangling images as well

andytson-inviqa committed:
Label multistage build sources so they can be deleted on destroy

In cases where docker image prune or --filter=since would have minor adverse effects on other projects, so it can even be run in local environments

andytson-inviqa committed:
Avoid image rm if no IMAGES to rm
